merlin/dtypes/base.py

```diff
@@ -0,0 +1,127 @@
+#
+# Copyright (c) 2022, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+
+from merlin.dtypes.registry import _dtype_registry
+
+
+class ElementType(Enum):
+    """
+    Merlin DType base types
+
+    Since a Merlin DType may describe a list, these are the either the types of
+    scalars or the types of list elements.
+    """
+
+    Bool = "bool"
+    Int = "int"
+    UInt = "uint"
+    Float = "float"
+    String = "string"
+    DateTime = "datetime"
+    Object = "object"
+    Unknown = "unknown"
+
+
+class ElementUnit(Enum):
+    """
+    Dtype units, used only for datetime types
+
+    Since a Merlin DType may describe a list, these are the either the units of
+    scalars or the units of list elements.
+    """
+
+    Year = "year"
+    Month = "month"
+    Day = "day"
+    Hour = "hour"
+    Minute = "minute"
+    Second = "second"
+    Millisecond = "millisecond"
+    Microsecond = "microsecond"
+    Nanosecond = "nanosecond"
+
+
+@dataclass(eq=True, frozen=True)
+class DType:
+    """
+    Merlin dtypes are objects of this dataclass
+    """
+
+    name: str
+    elemtype: ElementType
```

Comment:
one small question about naming. If users of this DType class are expected to use the element type, might be clearer here to use the full name for this `element_type` and same for the below two attrs?

Reply:
Yeah, that's a reasonable suggestion. These started out following the `itemtype` convention from `numpy`, which is how this ended up as `elemtype`. There's something to be said for brevity, but I agree it's clearer to use the full words for the attribute names.